File: 1146.snapshot-array.cpp
// @leet imports start
#include <bits/stdc++.h>
using namespace std;
// @leet imports end

// @leet start
class SnapshotArray {
  private:
    int snap_id = 0;
    vector<map<int, int>> arr;

  public:
    SnapshotArray(int length) : arr(length) {}

    void set(int index, int val) { arr[index][snap_id] = val; }

    int snap() { return snap_id++; }

    int get(int index, int snap_id) {
        // snaps = [0, 3]
        // to query at snap 2, if you take lower bound, you find 3
        // when you want snap 0
        auto it = arr[index].upper_bound(snap_id);
        if (it == arr[index].begin()) {
            return 0;
        }
        --it;
        return it->second;
    }
};

/**
 * Your SnapshotArray object will be instantiated and called as such:
 * SnapshotArray* obj = new SnapshotArray(length);
 * obj->set(index,val);
 * int param_2 = obj->snap();
 * int param_3 = obj->get(index,snap_id);
 */
// @leet end

File: 1146.snapshot-array.py
# @leet imports start
from typing import List, Optional

# @leet imports end

# @leet start
class SnapshotArray:

    def __init__(self, length: int):
        self.snap_id = 0
        self.arr = [{} for _ in range(length)]
        

    def set(self, index: int, val: int) -> None:
        self.arr[index][self.snap_id] = val
        

    def snap(self) -> int:
        self.snap_id += 1
        

    def get(self, index: int, snap_id: int) -> int:
        


# Your SnapshotArray object will be instantiated and called as such:
# obj = SnapshotArray(length)
# obj.set(index,val)
# param_2 = obj.snap()
# param_3 = obj.get(index,snap_id)
# @leet end
